dkg/nodesigs.go

```diff
@@ -49,33 +62,94 @@ func newNodeSigBcast(operatorAmt int, bcastComp *bcast.Component) nodeSigBcast {
 	return ret
 }
 
+// lockHash returns the lock hash byte array from lockHashCh.
+// Once the data has been received, it stays cached in lockHashData.
+func (n *nodeSigBcast) lockHash(ctx context.Context) ([]byte, error) {
+	n.lhLock.Lock()
+	defer n.lhLock.Unlock()
+
+	if len(n.lockHashData) != 0 {
+		return n.lockHashData, nil
+	}
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case n.lockHashData = <-n.lockHashCh:
+		return n.lockHashData, nil
+	}
+}
+
+// allSigs returns true if all the node signatures have been received.
+func (n *nodeSigBcast) allSigs() bool {
+	n.osLock.Lock()
+	defer n.osLock.Unlock()
+
+	for _, sig := range n.otherSigs {
+		if sig == nil {
+			return false
+		}
+	}
+
+	return true
+}
+
 // broadcastCallback is the default bcast.Callback for nodeSigBcast.
-func (n *nodeSigBcast) broadcastCallback(_ context.Context, _ peer.ID, _ string, msg proto.Message) error {
-	response, ok := msg.(*dkgpb.MsgNodeSig)
+func (n *nodeSigBcast) broadcastCallback(ctx context.Context, _ peer.ID, _ string, msg proto.Message) error {
+	nodeSig, ok := msg.(*dkgpb.MsgNodeSig)
 	if !ok {
 		return errors.New("invalid node sig type")
 	}
 
-	n.otherSigs <- response
+	sig := nodeSig.GetSignature()
+	msgPeerIdx := int(nodeSig.GetPeerIndex())
+
+	if (msgPeerIdx == n.nodeIdx.PeerIdx) || (msgPeerIdx < 0 || msgPeerIdx >= len(n.peers)) {
+		return errors.New("invalid peer index")
+	}
+
+	peerPubk, err := n.peers[msgPeerIdx].PublicKey()
+	if err != nil {
+		return errors.Wrap(err, "get peer public key")
+	}
+
+	lockHash, err := n.lockHash(ctx)
+	if err != nil {
+		return errors.Wrap(err, "lock hash wait")
+	}
+
+	verified, err := k1util.Verify(peerPubk, lockHash, sig[:len(sig)-1])
+	if err != nil {
+		return errors.Wrap(err, "verify signature")
+	} else if !verified {
+		return errors.New("invalid node signature")
+	}
+
+	n.osLock.Lock()
+	n.otherSigs[msgPeerIdx] = sig
+	n.osLock.Unlock()
 
 	return nil
 }
 
 // exchange exchanges K1 signatures over lock file hashes with the peers pointed by lh.bcastFunc.
 func (n *nodeSigBcast) exchange(
 	ctx context.Context,
-	lockHash []byte,
 	key *k1.PrivateKey,
-	nodeIdx cluster.NodeIdx,
+	lockHash []byte,
 ) ([][]byte, error) {
-	sig, err := k1util.Sign(key, lockHash)
+	localSig, err := k1util.Sign(key, lockHash)
 	if err != nil {
 		return nil, errors.Wrap(err, "k1 lock hash signature")
 	}
 
+	go func() {
+		n.lockHashCh <- lockHash
```

Comment:
ok, yeah, like that fact that you are passing lockhash here. Wonder if the channel is worth it still? Other option would be simple `n.setLockHash()` that uses the mutex and then a `n.awaitLockHash(ctx)` that just uses a for loop and a ticker to check if the value has been set. Not a fan of using channels AND mutexes, I would prefer to either or.

Reply:
i see your point, and in one of my original designs `exchange()` set `n.lockHash` directly so that the broadcast callbacks could use it, and another one where the caller had to set the lock hash manually

IMO this solution is fine, yeah we're using a lock + a channel but all in all the latter is only used once: it's doing its thing just once

Reply:
ok cool